src/channel_internal.c

```diff
@@ -279,148 +353,166 @@ static int dequeue_operation(THANDLE(CHANNEL_INTERNAL) channel_internal, THANDLE
+    if (on_data_available_cb != NULL)
     {
-        /*Codes_SRS_CHANNEL_INTERNAL_43_112: [ channel_internal_pull shall store the correlation_id, pull_callback and pull_context in the obtained operation. ]*/
+        /*Codes_SRS_CHANNEL_INTERNAL_43_112: [ channel_internal_pull shall store the correlation_id, on_data_available_cb and pull_context in the obtained operation. ]*/
         THANDLE_INITIALIZE(RC_STRING)(&channel_op->pull_correlation_id, pull_correlation_id);
-        channel_op->pull_callback = pull_callback;
+        channel_op->on_data_available_cb = on_data_available_cb;
         channel_op->pull_context = pull_context;
     }
-    else if (push_callback != NULL)
+    else if (on_data_consumed_cb != NULL)
     {
-        /*Codes_SRS_CHANNEL_INTERNAL_43_128: [ channel_internal_push shall store the correlation_id, push_callback, push_context and data in the obtained operation. ]*/
+        /*Codes_SRS_CHANNEL_INTERNAL_43_128: [ channel_internal_push shall store the correlation_id, on_data_consumed_cb, push_context and data in the obtained operation. ]*/
         THANDLE_INITIALIZE(RC_STRING)(&channel_op->push_correlation_id, push_correlation_id);
-        channel_op->push_callback = push_callback;
+        channel_op->on_data_consumed_cb = on_data_consumed_cb;
         channel_op->push_context = push_context;
         THANDLE_ASSIGN(RC_PTR)(&channel_op->data, data);
     }
-
-    /*Codes_SRS_CHANNEL_INTERNAL_43_114: [ channel_internal_pull shall set *out_op_pull to the THANDLE(ASYNC_OP) of the obtained operation. ] ]*/
-    /*Codes_SRS_CHANNEL_INTERNAL_43_130: [ channel_internal_push shall set *out_op_push to the THANDLE(ASYNC_OP) of the obtained operation. ]*/
-    THANDLE_INITIALIZE(ASYNC_OP)(out_op, channel_op->async_op);
-
+    THANDLE(ASYNC_OP) temp = NULL;
+    THANDLE_INITIALIZE(ASYNC_OP)(&temp, channel_op->async_op);
     /*Codes_SRS_CHANNEL_INTERNAL_43_113: [ channel_internal_pull shall call threadpool_schedule_work with execute_callbacks as work_function and the obtained operation as work_function_context. ]*/
     /*Codes_SRS_CHANNEL_INTERNAL_43_129: [ channel_internal_push shall call threadpool_schedule_work with execute_callbacks as work_function and the obtained operation as work_function_context. ]*/
     if (threadpool_schedule_work(channel_internal_ptr->threadpool, execute_callbacks, channel_op) != 0)
     {
         LogError("Failure in threadpool_schedule_work(channel_internal_ptr->threadpool=%p, execute_callbacks=%p, channel_op=%p)", channel_internal_ptr->threadpool, execute_callbacks, channel_op);
-        // undo the dequeue
-        THANDLE_ASSIGN(ASYNC_OP)(out_op, NULL);
-        if (pull_callback != NULL)
+        result = MU_FAILURE;
+        // For testing purposes, we must unset the callback
+        if (on_data_available_cb != NULL)
         {
-            channel_op->pull_callback = NULL;
-            channel_op->pull_context = NULL;
+            channel_op->on_data_available_cb = NULL;
         }
```

Comment:
seriously, what is going on here?  in the block above, you replace the old `on_data_availablre_cb` if a new one was passed in.  Here, if there's a failure, and you passed in a new callback, you set the stored one to null, which sounds like you're trying to un-do the replacement above.  I feel like I'm watching someone move shells around and I'm trying to figure out where the callback is being stored.

Reply:
[https://youtu.be/pXw7LYWNi5E?t=24](https://youtu.be/pXw7LYWNi5E?t=24)

Reply:
Removed.

Reply:
In the block above, the callback is being set ( not replaced) based on whether `dequeue_operation` was called from `channel_internal_push` or `channel_internal_pull`. 

`execute_callbacks` uses the state of whether the callbacks are `NULL` or non-`NULL` to decide the clean up that needs to be performed. 

In the test case where `threadpool_schedule_work` fails, the cleanup is performed immediately. Therefore the callback must be unset so that the cleanup doesn't happen two times. Does that make sense?